plugins/janus_streaming.c

```diff
@@ -3358,7 +3384,58 @@ void janus_streaming_incoming_rtcp(janus_plugin_session *handle, int video, char
 		JANUS_LOG(LOG_HUGE, "REMB for this PeerConnection: %"SCNu64"\n", bw);
 		/* TODO Use this somehow (e.g., notification towards application?) */
 	}
-	/* FIXME Maybe we should care about RTCP, but not now */
+	janus_streaming_session *session = (janus_streaming_session *)handle->plugin_handle;
+	if(!session || session->destroyed || session->stopping || !session->started || session->paused)
+		return;
+	janus_streaming_mountpoint *mp = (janus_streaming_mountpoint *)session->mountpoint;
+	if(mp->streaming_source != janus_streaming_source_rtp)
+		return;
+	janus_streaming_rtp_source *ss = (janus_streaming_rtp_source *)mp->source;
+
+	if ((video == 0) && (ss->audio_rtcp_fd > -1)) {
+		JANUS_LOG(LOG_VERB, "Got Audio RTCP SSRC %u\n",
+			janus_rtcp_get_sender_ssrc(buf, len));
+
+		janus_rtcp_fix_ssrc(NULL, buf, len, 1,
+			janus_rtcp_get_sender_ssrc(buf, len),
+			ss->audio_ssrc);
+
+		struct sockaddr_in dest_rtcp;
+		dest_rtcp.sin_family = AF_INET;
+		dest_rtcp.sin_port = htons(ss->audio_rtcp_port);
+		dest_rtcp.sin_addr.s_addr = ss->audio_mcast;
+		int sent = 0;
+
+		if (0 > (sent = sendto(ss->audio_rtcp_fd, buf, len, 0, &dest_rtcp, sizeof(dest_rtcp)))) {
+			JANUS_LOG(LOG_ERR, "Error sendto %d... %d (%s)\n",
+				ss->audio_rtcp_port, errno, strerror(errno));
+		} else {
+			JANUS_LOG(LOG_HUGE, "sent %d/%d bytes to %d\n",
+				sent, len, ss->audio_rtcp_port);
+		}
+	} else
+	if ((video != 0) && (ss->video_rtcp_fd > -1)) {
+		JANUS_LOG(LOG_VERB, "Got Video RTCP SSRC %u\n",
+			janus_rtcp_get_sender_ssrc(buf, len));
+
+		janus_rtcp_fix_ssrc(NULL, buf, len, 1,
+			janus_rtcp_get_sender_ssrc(buf, len),
+			ss->video_ssrc);
+
+		struct sockaddr_in dest_rtcp;
+		dest_rtcp.sin_family = AF_INET;
+		dest_rtcp.sin_port = htons(ss->video_rtcp_port);
+		dest_rtcp.sin_addr.s_addr = ss->video_mcast;
+		int sent = 0;
+
+		if (0 > (sent = sendto(ss->video_rtcp_fd, buf, len, 0, &dest_rtcp, sizeof(dest_rtcp)))) {
+			JANUS_LOG(LOG_ERR, "Error sendto %d... %d (%s)\n",
+				ss->video_rtcp_port, errno, strerror(errno));
+		} else {
+			JANUS_LOG(LOG_HUGE, "sent %d/%d bytes to %d\n",
+				sent, len, ss->video_rtcp_port);
+		}
+	} /* if */
```

Comment:
yes, of course. good point.. i've only been testing with one viewer, so my RTCP feedback case has been very simple. it seems i need to do play with this a little more, because i am seeing RR packets arrive at the source when this code is running.

Reply:
> i ultimately want to use REMB and RR to scale the source bitrate per the most bandwidth constrained viewer and as you've pointed out, blindly collapsing those messages from multiple viewers into a single SSRC will not work

But if the source of a Streaming mountpoint is a VideoRoom publisher from a Janus instance (the same or a remote one), these conflicting and fluctuating REMB messages will heavily confuse the browser: not to mention the HUGE overhead that REMB feedback coming from 300 viewers would cause, as REMB is quite chatty typically.

> this appears to be working due to whatever bug is allowing RR packets to be relayed to my source

Not a bug: RR and SR are passed to plugins, it's the other way that's cut down (if a plugin tries to push a RR/SR it's dumped by the core).

> NACK PLI seems like less of a problem

NACKs are stripped from the core before passing the RTCP packet to a plugin. I only meant PLI in my original answer: if you have 300 viewers, and all of them send a PLI in a time window of a second or so, no sense at all to forward 300 PLIs to the source; you can use some sort of timing info to decide when to send a new PLI (e.g., only send one per second at a max).

I understand different scenarios may have different requirements: if you have very few viewers it's not going to be an issue, when you have many it will; your source may be able to consume conflicting REMBs, other won't. I guess this means the behaviour here should ideally be customizable to some extent, so that one can use the one they feel works best for them.

Reply:
sorry if i wasn't clear--i am 100% in agreement about REMB.
i can't imagine a scenario where conflicting REMB messages can be useful to a source.

regarding PLI, i can see a case both for throttling these at the plugin as well as for letting the source handle them and it seems like it will be easy to add a configuration option to control it.

Reply:
Yeah, and with REMB it's not trivial either... for instance, if we decide to just forward the lowest REMB we receive, for how long do we stick to that value? Maybe it came from a temporary issue, or maybe the user left in the meanwhile, and it doesn't make sense to use such a low value anymore. I guess that one way to do that may be some form of window-based normalization: e.g., we track incoming REMBs for a second, and then forward the lowest one; for the next second, we do the same thing, and then send the updated REMB. It would mean sending periodic REMBs at a lower rate than how they're typically sent, but that might be a reasonable approach (maybe in a configurable way).

That said, thinking about it this is something that we might even decide to defer to a following effort: RTCP support here is optional anyway, so nothing happens if one doesn't enable it. If it's enabled, the considerations we made are a problem, but we can work on configuration options to modify the behaviour after this is merged. Meaning that I'm not saying you should take care of all this, it's something I can do as well, or others.

What's your feeling about this?